### ISSUE TITLE
Move limits HiOp optimizer

### ISSUE THREAD

**Issue by ricgross** [mfem/mfem#3639]
Hi, 

I have a brief question related to the HiOp optimizer, once again. 

@cnpetra and @vladotomov, is it possible to enforce move limits in-between iterations? 

I am trying to solve a highly nonlinear optimization problem, hence why I would like to enforce small design changes for the optimization itself to be as stable as possible. 

As far as I see, the only options I have available are to set the solution bounds, the absolute and the relative tolerance, none of which allow me to control the "step" between successive iterations.

In advance, thanks.
Best regards,
Riccardo

**Comment by vladotomov** [comment]
What does it mean `to enforce move limits`?

**Comment by ricgross** [comment]
Hi @vladotomov,

I would like to limit how much the design variables can change between successive iterations. I am trying to solve a topology optimization problem, which consists in minimizing the volume while subjected to a stress constraint.

However, I noticed that the optimizer is "quite aggressive", meaning that the change in volume between, let's say iteration 1 and 2, can be 20-30% of the total volume. 

Considering a problem such as the stress-constrained topology optimization one, where the design variables are the densities of each element (which can vary between 0 and 1), I would like a enforce a move limit, such that the current value of the density of a particular element cannot change more than a prescribed amount. 

If the density of a given element was 0.5 at iteration `i`,  I would like to limit the optimizer, such that in iteration `i+1`, the density (of the given element) can only change between, for example, 0.45 and 0.55 (if a move limit of 5% of the maximum density value is enforced). 

It can be compared to setting solution bounds at every iteration, with the difference that the solution bounds are not fixed, but they are dependent on the current design variable. 

Does it answer your question?

In advance, thanks.

Best regards,
Riccardo

**Comment by vladotomov** [comment]
Yes I understand. This is similar to line search in Newton methods where you scale the update.
@cnpetra can comment how this can be done in HiOp.

**Comment by cnpetra** [comment]
Moving limits/bounds are not supported by HiOp. There are algorithmic considerations why this is not supported. One way is do this is, as you alluded to, to continually restart the solver and move the density bounds as you find it necessary. I would let the optimizer take 5-6 iterations between restarts to build enough secant memory used to approximate the Hessian. 

Also, the initial aggressiveness is something expected with interior-point methods. You can reduce it by using milder reduction in the barrier parameters, for example set options `kappa_mu 0.6` and `theta_mu 1.1`.

**Comment by ricgross** [comment]
Hi @cnpetra,

Thank you for your reply.

How would you suggest to restart the solver code-wise? Are there any in-built methods which allow to perform such operation?

In advance, thanks.
Best regards,
Riccardo

**Comment by vladotomov** [comment]
> do you think it could be possible to implement the method `HiopOptimizationProblem::get_vars_info` (called at the beginning of the optimization problem) somewhere within the HiOp optimizer, such that it could read the input of the method `OptimizationProblem::SetSolutionBounds` after every iteration? In this way, the optimizer would be limited to only look for a design variable within the defined interval.

I think this has always been the case, see here https://github.com/mfem/mfem/blob/6dde5832bc819c9396ac983542dff44d4f5b6e5e/linalg/hiop.cpp#L43
